build_tools/github_actions/runner/instance_deleter/main.py

```diff
@@ -0,0 +1,211 @@
+# Copyright 2022 The IREE Authors
+#
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+"""A Cloud Functions proxy enabling GCE VMs in a Managed Instance Group to delete themselves.
+
+GCE Managed instance groups don't have any good way to handle autoscaling for
+long-running workloads. With the autoscaler configured to scale in, instances
+get only 90 seconds warning to shut down. So we set the autoscaler to only scale
+out and have the VMs tear themselves down when they're down with their work.
+This is the approach suggested by the managed instance group team:
+
+https://drive.google.com/file/d/1XlwxF_0T7pUnbzhL5ePDoW-Q3GAaLO11
+
+But anything that brings down the VM other than a delete call to the instance
+group manager API makes the VM get considered "unhealthy", which means it gets
+recreated in exactly the same configuration, regardless of any update or
+autoscaling settings. Making the correct API call requires broad permissions on
+the instance group manager, which we don't want to give the VMs. To scope
+permissions to individual instances, this proxy service makes use of instance
+identity tokens to allow an instance to make a call only to delete itself.
+
+See
+https://cloud.google.com/compute/docs/instances/verifying-instance-identity
+
+This makes use of the GCP Cloud Functions serverless offering. It's another
+level of abstraction on top of Cloud Run, where you don't even need to create your
+own docker container. For local development:
+
+  functions-framework --target=delete_self
+  curl -X DELETE -v --header "Authorization: Bearer $(cat /tmp/token.txt)" localhost:8080
+
+You'll need to get a token that corresponds to an actual instance though or
+you'll get an error:
+
+  gcloud compute ssh github-runner-testing-presubmit-cpu-us-west1-h58j \
+    --user-output-enabled=false \
+    --command "curl -sSfL \
+        -H 'Metadata-Flavor: Google' \
+        'http://metadata/computeMetadata/v1/instance/service-accounts/default/identity?audience=localhost&format=full'" \
+    > /tmp/token.txt
+
+To deploy:
+  gcloud functions deploy instance-self-deleter \
+    --gen2 \
+    --runtime=python310 \
+    --region=us-central1 \
+    --source=. \
+    --entry-point=delete_self \
+    --trigger-http
+
+See https://cloud.google.com/functions/docs for more details.
+"""
+
+import flask
+import functions_framework
+import requests
+from google.api_core import exceptions
```

Comment:
nit: I had a little trouble when seeing `exceptions.NotFound`. Maybe import it as something like `api_exceptions`?

Reply:
Ah, right pytype got upset when I did complete imports because it doesn't know about the attributes on modules we don't have installed. I've turned off that particular error since we're already disabling import-error for the same reason.